doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+  
+  Triggering the delete problem:
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString (copiedKey) == keyString (originalKey));
+  assert (keyName (copiedKey) == keyName (originalKey));
+
+  keyDel (originalKey);
+  
+  keyString (copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
+  keyName (copiedKey);   // Error! Original name has been deleted.
+  ```
+ 
+  Triggering the update problem:
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString(copiedKey) == keyString(originalKey));
+  keySetString (originalKey, "new value!");
+  
+  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
+  ```
+
+  The same problems in principle exist for `mmapstorage`, however this is just a storage plugin and the user does not have any API to create COW keys themselves.
+  We can still let users access the flag `ELEKTRA_CP_COW`, we just need to clearly document what is forbidden.
+  Maybe set the `KEY_FLAG_RO_VALUE` on the original key, so that the API itself detects the error. 
+  There is, however, no flag for `keyDel` that we could set.
+
+### Data restrictions
+
+@kodebach wrote:
+
+> Make all the keys returned by kdbGet completely read-only.
+> To change the data you need to append an entirely new key to replace the existing one.
+> Then we just need to keep a shallow copy internally.
+
+### API restrictions
+
+@kodebach wrote:
+
+> Change the API and remove KeySet from kdbGet and kdbSet also option 4 in [the operation sequences decision](../0_drafts/operation_sequences.md).
+> If the keyset is owned by the KDB handle, it should not be as big surprise, if there is extra data in there.
+> I certainly wouldn't try to asset anything on the contents of a KeySet that I don't own directly, unless the condition is explicitly documented somewhere.
+
+@markus2330 wrote:
+
+> I disagree, it is actually the same kind of surprise for "More Keys".
+> Only the "Fewer Keys" would get fixed.
+
+### Full-blown copy-on-write implementation
+
+Make Elektra's `Key` and `KeySet` data structures copy-on-write.
+This requires some major refactoring of code within `libelektra-core`.
+Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
+The `mmapstorage` plugin will need a major refactoring.
+
+#### Changes to `Key`
+
+For the `Key`, we need to extract everything for the data and name into their own structs.
+This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.
+
+```c
+struct _KeyData {
+    union {
+        char * c;
+        void * v;
+    } data;
+
+    size_t dataSize;
+
+    uint16_t refs;
+};
+
+struct _KeyName {
+    char * key;
+    size_t keySize;
```

Comment:
Actually we want to get rid of this, see ./2_in_progress/store_name.md

Reply:
Okay. But what does "in progress" mean in that case? Is somebody actively working on it?

Reply:
Yes, @mpranj worked on it.

Reply:
AFAIK there has been zero progress on removing the key names. Also "In Progress" is _before_ "Decided" so that's extra confusing and more reason why I suggested to remove "In Progress" in #4639.